### PR TITLE
LMConsolidator // Implementation.

### LINKED ISSUE #24: 待辦：主動統整使用者語彙檔案內容、且清理可能存在的重複條目。
輸入法目前有個函數是用來載入使用者語彙的。
這個函數需要改進：在正式載入之前，對該檔案進行就地處理。
處理內容順序如下：

一、檢索所有的全形空格與 tab，如果有的話則一律取代為半形空格。
二、檢索所有的連續的半形空格，如果有的話則一律取代為單個半形空格。
三、檢索所有存在於行末的半形空隔，如果有的話則一律將其清除。
四、檢索任何可能重複的行，如果有的話則僅保留其中的一份。

Commit message 定義如下：`Author: UserPhraseLM // Deduplicate lines when loading.`

### PR DESCRIPTION
**Pull request by ShikiSuen**
We don't prefer coward methods like only making LM modules "tolerant".
We actively consolidate user-editable files to fix common user-generated mistakes and duplicated entries.

**All commit titles are of this format:** `LMConsolidator // The Quick Brown Fox...`.

Dev Goals for the LMConsolidator (tick if really finished):

- [x] Deprecate the EOF fix in FastLM. (I forgot to close the `zfd` ifstream.)
- [x] Implement the LMConsolidator module in lieu of the previously-implemented EOF fix codeblocks.
- [x] Implement the de-duplicator function to the LMConsolidator.
- [x] Implement the format-consolidator function to the LMConsolidator.
- [x] Unit tests.

All possible incorrect author information will be fix-amended after the unit tests prior to the final merge.

### CODE REVIEW

**Comment by ShikiSuen** [comment]
The Format Consolidator has a problem.
I commented those lines (in the LMConsolidator module) regarding the use of regex.
As soon as they get uncommented and implemented in the context, the IME gets choked and cannot type anything.

@hirakujira could you please help me take a look at it?

**Comment by ShikiSuen** [comment]
看來需要引入這個模組：https://github.com/alllexx88/pcrscpp

屆時用的 SED："s/[\\h]*/ /g;s/[\\s]*\\n/\\n/g"

**Comment by ShikiSuen** [comment]
Just decided to use Swift to finish the regex replacement part.
C++ simply doesn't work without installing Boost library which is too big to be implemented to vChewing.

**Comment by ShikiSuen** [comment]
Deprecating Zonble's on-write EOF fixer.

Zonble thinks that his idea is superior than mine. However, his design, by putting an EOF fixer into the writeUserPhrase section, is totally meaningless: EOF fixer is for fixing manually-introduced EOF faults since automatically-inserted UserPhrase entries never generates EOF problems.

**Comment by ShikiSuen** [comment]
The Format Consolidator is finally finished. It does the following per-line cleanup in sequence:

1. Omit blank lines.
2. Automatically replaces CJK whitespaces to ASCII space.
3. Automatically replaces every consecutive RegEx space section (type `\s+`) to a single ASCII space.
4. Remove leading and trailing spaces from each line.
5. Omit this line if this line becomes blank after the cleaning processes performed above.

File: Source/Engine/LanguageModel/LMConsolidator.mm [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+        if (line.empty()) {
+            continue;
+        }
+        auto res = container.insert(line);
+        if (res.second) {
+            index[&(*res.first)] = idx;
+            mp[idx] = {res.first, 1};
+            idx++;
+        } else {
+            mp[index[&(*res.first)]].second += 1;
```
Comment (hirakujira) [inline Source/Engine/LanguageModel/LMConsolidator.mm]: 後面沒有要顯示次數，所以這邊不需要計算次數

File: Source/Engine/LanguageModel/LMConsolidator.mm [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+    string line;
+    int idx = 0;
+    while (getline(zfdDeduplicateIncomingStream, line)) {
+        if (line.empty()) {
+            continue;
+        }
+        auto res = container.insert(line);
+        if (res.second) {
+            index[&(*res.first)] = idx;
+            mp[idx] = {res.first, 1};
```
Comment (hirakujira) [inline Source/Engine/LanguageModel/LMConsolidator.mm]: 因為沒有要計算次數，所以這邊可以更簡化，應該不需要用到 mp 這個 map，可以直接用 vector 存資料就好。index 這個 map 看起來似乎也不需要。

**Comment by ShikiSuen** [comment]
I rewrote the deduplicator module (rebased & replaced the old one).
Now, I pushed a new commit merging the deduplicator module and the format-consolidator module into a new single module: content-consolidator.

**Comment by ShikiSuen** [comment]
Fixed a typo in a commit (not the latest one).
I guess that's the end for tonight.

**Pull request merged**